tests/t0115_gateway_dir_listing_test.go

```diff
@@ -0,0 +1,364 @@
+			//   curl -sD - --resolve $DIR_HOSTNAME:$GWAY_PORT:127.0.0.1 http://$DIR_HOSTNAME:$GWAY_PORT/ą/ę > list_response &&
+			//   test_should_contain "HTTP/1.1 301 Moved Permanently" list_response &&
+			//   test_should_contain "Location: /%c4%85/%c4%99/" list_response
+			// '
+			{
+				Name: "redirect dir listing to URL with trailing slash",
+				Request: Request().DoNotFollowRedirects().URL(
+					"{{scheme}}://{{cid}}.ipfs.{{host}}/ą/ę",
+					u.Scheme,
+					root.Cid(),
+					u.Host,
+				),
+				Response: Expect().
+					Status(301).
+					Headers(
+						Header("Location").Equals(`/%c4%85/%c4%99/`),
+					),
+			},
+			// test_expect_success "subdomain gw: Etag should be present" '
+			//   curl -sD - --resolve $DIR_HOSTNAME:$GWAY_PORT:127.0.0.1 http://$DIR_HOSTNAME:$GWAY_PORT/ą/ę/ > list_response &&
+			//   test_should_contain "Index of" list_response &&
+			//   test_should_contain "Etag: \"DirIndex-" list_response
+			// '
+			// test_expect_success "subdomain gw: backlink on subdirectory should point at parent directory" '
+			//   test_should_contain "<a href=\"/%C4%85/%C4%99/..\">..</a>" list_response
+			// '
+			// test_expect_success "subdomain gw: breadcrumbs should leverage path-based router mounted on the parent domain" '
+			//   test_should_contain "/ipfs/<a href=\"//localhost:$GWAY_PORT/ipfs/$DIR_CID\">$DIR_CID</a>/<a href=\"//localhost:$GWAY_PORT/ipfs/$DIR_CID/%C4%85\">ą</a>/<a href=\"//localhost:$GWAY_PORT/ipfs/$DIR_CID/%C4%85/%C4%99\">ę</a>" list_response
+			// '
+			// test_expect_success "subdomain gw: name column should be a link to content root mounted at subdomain origin" '
+			//   test_should_contain "<a href=\"/%C4%85/%C4%99/file-%C5%BA%C5%82.txt\">file-źł.txt</a>" list_response
+			// '
+			// test_expect_success "subdomain gw: hash column should be a CID link to path router with filename param" '
+			//   test_should_contain "<a class=\"ipfs-hash\" translate=\"no\" href=\"//localhost:$GWAY_PORT/ipfs/$FILE_CID?filename=file-%25C5%25BA%25C5%2582.txt\">" list_response
+			// '
+			{
+				Name: "Regular dir listing",
+				Request: Request().DoNotFollowRedirects().URL(
+					"{{scheme}}://{{cid}}.ipfs.{{host}}/ą/ę/",
+					u.Scheme,
+					root.Cid(),
+					u.Host,
+				),
+				Response: Expect().
+					Headers(
+						Header("Etag").Contains(`"DirIndex-`),
+					).BodyWithHint(`
+					- backlink on subdirectory should point at parent directory
+					- breadcrumbs should leverage path-based router mounted on the parent domain
+					- name column should be a link to content root mounted at subdomain origin
+					`,
+					And(
+						Contains("Index of"),
+						Contains(
+							`<a href="/%C4%85/%C4%99/..">..</a>`,
+						),
+						Contains(
+							`/ipfs/<a href="//{{host}}/ipfs/{{cid}}">{{cid}}</a>/<a href="//{{host}}/ipfs/{{cid}}/%C4%85">ą</a>/<a href="//{{host}}/ipfs/{{cid}}/%C4%85/%C4%99">ę</a>`,
+							u.Host, // TODO: That looks wrong, we link without considering the subdomain
```

Comment:
I asked @lidel directly:
we do not redirect to the subdomain URL, because cidv0 would break with normalization,
it's easier to link to the /ipfs/cid for all cases and let the gateway deal with subdomain'ing.